clang-tools-extra/clang-tidy/utils/UseRangesCheck.cpp

```diff
@@ -0,0 +1,253 @@
+      std::vector<ast_matchers::internal::DynTypedMatcher> Matchers;
+      for (const auto &ArgPair : Signature) {
+        Matchers.push_back(makeMatcherPair(getFullPrefix(Signature), ArgPair));
+      }
+      TotalMatchers.push_back(
+          ast_matchers::internal::DynTypedMatcher::constructVariadic(
+              ast_matchers::internal::DynTypedMatcher::VO_AllOf,
+              ASTNodeKind::getFromNodeKind<CallExpr>(), std::move(Matchers)));
+    }
+    Finder->addMatcher(
+        callExpr(
+            callee(functionDecl(hasAnyName(std::move(Names))).bind(FuncDecl)),
+            ast_matchers::internal::DynTypedMatcher::constructVariadic(
+                ast_matchers::internal::DynTypedMatcher::VO_AnyOf,
+                ASTNodeKind::getFromNodeKind<CallExpr>(),
+                std::move(TotalMatchers))
+                .convertTo<CallExpr>()),
+        this);
+  }
+}
+
+// Returns the proper token based end location of \p E.
+static SourceLocation exprLocEnd(const Expr *E, const ASTContext &Context) {
+  return Lexer::getLocForEndOfToken(
+      E->getEndLoc(), 0, Context.getSourceManager(), Context.getLangOpts());
+}
+
+static void removeFunctionArgs(const CallExpr &Call, ArrayRef<unsigned> Indexes,
+                               llvm::SmallVectorImpl<FixItHint> &Output,
+                               const ASTContext &Ctx) {
+  llvm::SmallVector<unsigned> Sorted(Indexes);
+  // Keep track of commas removed
+  llvm::SmallBitVector Commas(Call.getNumArgs());
+  // The first comma is actually the '(' which we can't remove
+  Commas[0] = true;
+  llvm::sort(Sorted);
+  for (auto Index : Sorted) {
+    const auto *Arg = Call.getArg(Index);
+    if (Commas[Index]) {
+      if (Index >= Commas.size()) {
+        Output.push_back(FixItHint::CreateRemoval(Arg->getSourceRange()));
+      } else {
+        // Remove the next comma
+        Commas[Index + 1] = true;
+        Output.push_back(
+            FixItHint::CreateRemoval(CharSourceRange::getTokenRange(
+                {Arg->getBeginLoc(),
+                 exprLocEnd(Arg, Ctx).getLocWithOffset(1)})));
+      }
+    } else {
+      Output.push_back(FixItHint::CreateRemoval(CharSourceRange::getTokenRange(
+          Arg->getBeginLoc().getLocWithOffset(-1), Arg->getEndLoc())));
+      Commas[Index] = true;
+    }
+  }
+}
+
+void UseRangesCheck::check(const MatchFinder::MatchResult &Result) {
+  const auto *Function = Result.Nodes.getNodeAs<FunctionDecl>(FuncDecl);
+  auto QN = "::" + Function->getQualifiedNameAsString();
+  auto Iter = Replaces.find(QN);
+  assert(Iter != Replaces.end());
+  SmallString<64> Buffer;
+  for (const auto &Signature : Iter->getValue()->getReplacementSignatures()) {
+    Buffer.assign({BoundCall, getFullPrefix(Signature)});
+    const auto *Call = Result.Nodes.getNodeAs<CallExpr>(Buffer);
+    if (!Call)
+      continue;
+    auto Diag =
+        diag(Call->getBeginLoc(), "Use a ranges version of this algorithm");
```

Comment:
"Use a ranges version of this algorithm" -> "use a ranges version of this algorithm"